### PR TITLE
add support xxl-job

### PR DESCRIPTION
**Pull request by IAMTJW**
Please answer these questions before submitting pull request

- Why submit this pull request?
- [ ] Bug fix
- [ ] New feature provided
- [x] Improve performance

- Related issues

___
### Bug fix
- Bug description.

- How to fix?

___
### New feature or improvement
- Describe the details and related test reports.
add skywalking support xxl-job like elastic-job

### CODE REVIEW

**Comment by wu-sheng** [comment]
You need to submit test case to here https://github.com/SkyAPMTest/agent-auto-integration-testcases 

You could find the document there too.

**Comment by zhaoyuguang** [comment]
I suggest implementing the `cron expression job` on the tracking the `xxl server` to deliver the task to the client.

**Comment by IAMTJW** [comment]
> You need to submit test case to here https://github.com/SkyAPMTest/agent-auto-integration-testcases
> 
> You could find the document there too.

the xxl-job like elastic-job and the test case may be difficult to make

**Comment by IAMTJW** [comment]
> I suggest implementing the `cron expression job` on the tracking the `xxl server` to deliver the task to the client.

the server to client use http and the client run job in new thread and it is not work

**Comment by wu-sheng** [comment]
> the xxl-job like elastic-job and the test case may be difficult to make

Could you use a HTTP request to trigger a job execution? If YES, then test still works.

**Comment by wu-sheng** [comment]
When you are ready, please ping me or @ascrutae

**Comment by IAMTJW** [comment]
> When you are ready, please ping me or @ascrutae

I am ready and What should I do next

**Comment by wu-sheng** [comment]
@ascrutae Please review the plugin and test case(https://github.com/SkyAPMTest/agent-auto-integration-testcases/pull/62), if everything looks right, please submit the integration test in our INFRA.

**Comment by wu-sheng** [comment]
@IAMTJW Please update this PR to fix CI.

**Comment by wu-sheng** [comment]
You need to solve conflict. Another plugin merged today.

File: apm-sniffer/apm-sdk-plugin/xxl-job-plugin/src/main/java/org/apache/skywalking/apm/plugin/xxljob/JobHandlerInterceptor.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        } else {
+            operateName = ComponentsDefine.XXL_JOB.getName();
+        }
+        ShardingUtil.ShardingVO shardingVO = ShardingUtil.getShardingVo();
+        if (shardingVO != null) {
+
+            operateName = operateName + "_" + shardingVO.getIndex();
+        }
+        AbstractSpan span;
+        span = ContextManager.createEntrySpan(operateName, null);
```
Comment (wu-sheng) [inline apm-sniffer/apm-sdk-plugin/xxl-job-plugin/src/main/java/org/apache/skywalking/apm/plugin/xxljob/JobHandlerInterceptor.java]: Create an Entry span but without ContextCarrier. Could you tell me why?

Comment (wu-sheng) [reply apm-sniffer/apm-sdk-plugin/xxl-job-plugin/src/main/java/org/apache/skywalking/apm/plugin/xxljob/JobHandlerInterceptor.java]: cc @ascrutae

Comment (wu-sheng) [reply apm-sniffer/apm-sdk-plugin/xxl-job-plugin/src/main/java/org/apache/skywalking/apm/plugin/xxljob/JobHandlerInterceptor.java]: Why do you create entry span? Entry span means RPC related, but I can't find another exit span somewhere. From the codes I saw, this is only a local span to group following spans.

Comment (wu-sheng) [reply apm-sniffer/apm-sdk-plugin/xxl-job-plugin/src/main/java/org/apache/skywalking/apm/plugin/xxljob/JobHandlerInterceptor.java]: > I suggest implementing the cron expression job on the tracking the xxl server to deliver the task to the client.

Noticed @zhaoyuguang comments. If you have server side, you could choose to implement server side plugin, with ContextCarrier propagation. Or just create a local span, without propagation.

But you need to define and be clear, what your plugin for, and how it should work.

Comment (IAMTJW) [reply apm-sniffer/apm-sdk-plugin/xxl-job-plugin/src/main/java/org/apache/skywalking/apm/plugin/xxljob/JobHandlerInterceptor.java]: the server side is web server，and the server side call client by  http （RPC），and I use the createEntrySpan

Comment (IAMTJW) [reply apm-sniffer/apm-sdk-plugin/xxl-job-plugin/src/main/java/org/apache/skywalking/apm/plugin/xxljob/JobHandlerInterceptor.java]: if the server side is deployed without skywaling , the client maybe use the  createLocalSpan like elastic-job

Comment (wu-sheng) [reply apm-sniffer/apm-sdk-plugin/xxl-job-plugin/src/main/java/org/apache/skywalking/apm/plugin/xxljob/JobHandlerInterceptor.java]: If this goes through HTTP, which plugin acts a web container? I don't think you handler should create entry span in either case. Entry span means the incoming call, but in the current case, I think your plugin has nothing related to RPC.

Comment (wu-sheng) [reply apm-sniffer/apm-sdk-plugin/xxl-job-plugin/src/main/java/org/apache/skywalking/apm/plugin/xxljob/JobHandlerInterceptor.java]: > if the server side is deployed without skywaling , the client maybe use the createLocalSpan like elastic-job

You are creating entry span in your codes.

**Comment by wu-sheng** [comment]
Basically, I need you to explain the plugin design. I am a little confused about what you are trying to do here.
1. When RPC happens, which plugins work for tracing? Such as, Tomcat? HttpClient? Or something else?
1. If the local span of handler is the only thing created, then what is the key point to add this plugin? Is this plugin just playing the simplest role, which is adding a span in an existing trace, but not related to any metrics of service or topology map.

**Comment by IAMTJW** [comment]
> Basically, I need you to explain the plugin design. I am a little confused about what you are trying to do here.
> 
> 1. When RPC happens, which plugins work for tracing? Such as, Tomcat? HttpClient? Or something else?
> 2. If the local span of handler is the only thing created, then what is the key point to add this plugin? Is this plugin just playing the simplest role, which is adding a span in an existing trace, but not related to any metrics of service or topology map.



> Basically, I need you to explain the plugin design. I am a little confused about what you are trying to do here.
> 
> 1. When RPC happens, which plugins work for tracing? Such as, Tomcat? HttpClient? Or something else?
> 2. If the local span of handler is the only thing created, then what is the key point to add this plugin? Is this plugin just playing the simplest role, which is adding a span in an existing trace, but not related to any metrics of service or topology map.

1、想通过这个插件知道调用链发起是由 xxl-job 服务产生的调用流程，类似elastic-job
2、xxl-job 有服务的 admin 和 客户端两部分组成，任务的调用时通过 服务端 通过 JettyClient（http ）rpc调用来触发 客户端执行的。
3、xxl-job服务端 http 调用 客户端的过程中会触发与web相关的拦截器，但由于 xxl-job 客户端线程池执行服务，无法标记出调用链是由 xxl-job 发起的，并且xxl-job 服务端在部署的时候无法保证部署使用了 skywaling

插件出发点：类似 elastic-job，能标记出调用链是由 xxl-job 发起的

**Comment by wu-sheng** [comment]
You need to translate the reply to English.

**Comment by IAMTJW** [comment]
> Basically, I need you to explain the plugin design. I am a little confused about what you are trying to do here.
> 
> 1. When RPC happens, which plugins work for tracing? Such as, Tomcat? HttpClient? Or something else?
> 2. If the local span of handler is the only thing created, then what is the key point to add this plugin? Is this plugin just playing the simplest role, which is adding a span in an existing trace, but not related to any metrics of service or topology map.



> Basically, I need you to explain the plugin design. I am a little confused about what you are trying to do here.
> 
> 1. When RPC happens, which plugins work for tracing? Such as, Tomcat? HttpClient? Or something else?
> 2. If the local span of handler is the only thing created, then what is the key point to add this plugin? Is this plugin just playing the simplest role, which is adding a span in an existing trace, but not related to any metrics of service or topology map.

1、When RPC happens, the jetty plugin work for tracing.
2、when the xxl-job client is called , because of the client execute job with thread pool and intercept IJobHandler to know the tracing from xxl-job like elastic-job

**Comment by wu-sheng** [comment]
> 1、When RPC happens, the jetty plugin work for tracing.
2、when the xxl-job client is called , because of the client execute job with thread pool and intercept IJobHandler to know the tracing from xxl-job like elastic-job

**Comment by wu-sheng** [comment]
> 1、When RPC happens, the jetty plugin work for tracing.

If so, then no entry span required.

> 2、when the xxl-job client is called , because of the client execute job with thread pool and intercept IJobHandler to know the tracing from xxl-job like elastic-job

If there is a thread pool in there, you just added a local span, the trace will still break. The right thing is you make the trace continues by using `Context#capture`, `Context#continued` APIs.  Then you know the trace starts with a xxl-job server, and including all spans of your codes.

**Comment by wu-sheng** [comment]
Without update in one week, will close soon

File: apm-sniffer/apm-sdk-plugin/xxl-job-plugin/src/main/java/org/apache/skywalking/apm/plugin/xxljob/ExecutorBizInterceptor.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    @Override
+    public void beforeMethod(EnhancedInstance objInst, Method method, Object[] allArguments, Class<?>[] argumentsTypes, MethodInterceptResult result) throws Throwable {
+
+        TriggerParam triggerParam = (TriggerParam) allArguments[0];
+        String operateName = triggerParam.getExecutorHandler();
+        if (triggerParam.getBroadcastIndex() != 0) {
+            operateName += "-" + triggerParam.getBroadcastIndex();
+        }
+        ContextSnapshot contextSnapshot = ContextManager.capture();
+        contextSnapshot.setEntryOperationName(operateName);
```
Comment (wu-sheng) [inline apm-sniffer/apm-sdk-plugin/xxl-job-plugin/src/main/java/org/apache/skywalking/apm/plugin/xxljob/ExecutorBizInterceptor.java]: Override entry operation name? Why?

**Comment by wu-sheng** [comment]
Be advised, RESTEasy plugin merged, conflict needs to be solved.

**Comment by IAMTJW** [comment]
> I am still not sure the result of this plugin, could you provide
> 
> 1. Topology and trace when xxl server and xxl client under SkyWalking monitoring
> 2. Topology and trace when only xxl client under SkyWalking monitoring

1、Sorry，the xxl-job server call client with JettyHttpClient （and the new version remove it）and the skywalking not support jetty plugin.
2、If the xxl server not under the SkyWalking monitoring ContextManager.capture() may be NullPointerException

**Comment by wu-sheng** [comment]
> 2、If the xxl server not under the SkyWalking monitoring ContextManager.capture() may be NullPointerException

This is not the factor. I think you misunderstand the API usage.

**Pull request closed without merging**